pkg/repro/repro.go

```diff
@@ -168,46 +243,212 @@ func (ctx *context) reproExtractProg(entries []*prog.LogEntry) (*Result, error)
 		Repro:      true,
 	}
 
-	// Execute the suspected programs.
-	// We first try to execute each program for 10 seconds, that should detect simple crashes
-	// (i.e. no races and no hangs). Then we execute each program for 5 minutes
-	// to catch races and hangs. Note that the max duration must be larger than
-	// hang/no output detection duration in vm.MonitorExecution, which is currently set to 3 mins.
-	// Programs are executed in reverse order, usually the last program is the guilty one.
-	durations := []time.Duration{10 * time.Second, 5 * time.Minute}
-	suspected := [][]*prog.LogEntry{reverseEntries(entries), reverseEntries(lastEntries)}
-	var res *Result
-	for i, dur := range durations {
-		for _, ent := range suspected[i] {
-			opts.Fault = ent.Fault
-			opts.FaultCall = ent.FaultCall
-			opts.FaultNth = ent.FaultNth
-			if opts.FaultCall < 0 || opts.FaultCall >= len(ent.P.Calls) {
-				opts.FaultCall = len(ent.P.Calls) - 1
+	for _, ent := range entries {
+		opts.Fault = ent.Fault
+		opts.FaultCall = ent.FaultCall
+		opts.FaultNth = ent.FaultNth
+		if opts.FaultCall < 0 || opts.FaultCall >= len(ent.P.Calls) {
+			opts.FaultCall = len(ent.P.Calls) - 1
+		}
+		crashed, err := ctx.testProg(ent.P, duration, opts)
+		if err != nil {
+			return nil, err
+		}
+		if crashed {
+			res := &Result{
+				Prog:     ent.P,
+				Duration: duration * 3 / 2,
+				Opts:     opts,
 			}
-			crashed, err := ctx.testProg(ent.P, dur, opts)
+			return res, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (ctx *context) reproExtractProgBisect(entries []*prog.LogEntry, baseDuration time.Duration) (*Result, error) {
+	Logf(3, "reproducing crash '%v': bisect: bisecting %d programs", ctx.crashDesc, len(entries))
+
+	opts := csource.Options{
+		Threaded:   true,
+		Collide:    true,
+		Repeat:     true,
+		Procs:      ctx.cfg.Procs,
+		Sandbox:    ctx.cfg.Sandbox,
+		EnableTun:  true,
+		UseTmpDir:  true,
+		HandleSegv: true,
+		WaitRepeat: true,
+		Debug:      true,
+		Repro:      true,
+	}
+
+	duration := func(entries int) time.Duration {
+		return baseDuration + time.Duration((entries/4))*time.Second
+	}
+
+	// Check that executing the whole log results in a crash.
+	Logf(3, "reproducing crash '%v': bisect: executing all %d programs", ctx.crashDesc, len(entries))
+	crashed, err := ctx.testProgs(entries, duration(len(entries)), opts)
+	if err != nil {
+		return nil, err
+	}
+	if !crashed {
+		Logf(3, "reproducing crash '%v': bisect: didn't crash", ctx.crashDesc)
+		return nil, nil
+	}
+
+	compose := func(guilty1, guilty2 [][]*prog.LogEntry, chunk1, chunk2 []*prog.LogEntry) []*prog.LogEntry {
```

Comment:
I don't see that you ever pass both chunk1 and chunk2. Should we remove the second chunk argument?